core/src/main/java/org/apache/iceberg/io/BaseTaskWriter.java

```diff
@@ -151,9 +146,14 @@ private void internalPosDelete(StructLike key) {
      * @param row the given row to delete.
      */
     public void delete(T row) throws IOException {
-      internalPosDelete(structProjection.wrap(asStructLike(row)));
+      StructProjection key = structProjection.wrap(asStructLike(row));
+      PathOffset previous = insertedRowMap.get(key);
 
-      eqDeleteWriter.write(row);
+      if (previous != null){
+        posDeleteWriter.delete(previous.path, previous.rowOffset, null);
+      } else {
+        eqDeleteWriter.write(row);
+      }
```

Comment:
Overall the logic change in this function looks reasonable to me.

Reply:
I noticed a problem with this: it doesn't remove the `PathOffset` from `insertedRowMap` the way `internalPosDelete` did. See my comment above.